tests/cache/test_mesi_controller.py

```diff
@@ -1,116 +1,479 @@
-        self.assertIsNone(line)
-
-    # TODO: Add more detailed MESI tests:
-    # - Test snoop handling
-    # - Test victim line handling
-    # - Test all possible state transitions
-    # - Test multiple lines in same set with different states
-    # - Test edge cases and error conditions
-
-
-if __name__ == "__main__":
-    unittest.main()
+    ############################################################################################################
+    # Processor request tests
+
+    def test_processor_read_invalid_and_nohit(self):
+        """
+        Test handling a processor read request for a cacheline in INVALID state and NOHIT snoop result
+        """
+        address = 0x1000
+        self.mock_get_snoop.return_value = SnoopResult.NOHIT
+
+        # Call the function under test
+        next_state = self.controller.handle_processor_request(MESIState.INVALID, address, is_processor_write=False)
+
+        # Check that the bus operation was called with the correct arguments
+        self.mock_bus_op.assert_called_with(BusOp.READ, 0x1000)
+
+        # Check that the snoop result was checked
+        self.mock_get_snoop.assert_called_with(0x1000)
+
+        # Assert state transition: INVALID -> EXCLUSIVE
+        self.assertEqual(next_state, MESIState.EXCLUSIVE)
+
+
+    def test_processor_read_invalid_and_hit(self):
+        """
+        Test handling a processor read request for a cacheline in INVALID state and HIT snoop result
+        """
+        address = 0x1000
+        self.mock_get_snoop.return_value = SnoopResult.HIT
+
+        # Call the function under test
+        next_state = self.controller.handle_processor_request(MESIState.INVALID, address, is_processor_write=False)
+
+        # Check that the bus operation was called with the correct arguments
+        self.mock_bus_op.assert_called_with(BusOp.READ, 0x1000)
+
+        # Check that the snoop result was checked
+        self.mock_get_snoop.assert_called_with(0x1000)
+
+        # Assert state transition: INVALID -> SHARED
+        self.assertEqual(next_state, MESIState.SHARED)
+
+    def test_processor_write_invalid_and_nohit(self):
+        """
+        Test handling a processor write request for a cacheline in INVALID state and NOHIT snoop result
+        """
+        address = 0x1000
+        self.mock_get_snoop.return_value = SnoopResult.NOHIT
+
+        # Call the function under test
+        next_state = self.controller.handle_processor_request(MESIState.INVALID, address, is_processor_write=True)
+
+        # Check that the bus operation was called with the correct arguments
+        self.mock_bus_op.assert_called_with(BusOp.RWIM, 0x1000)
+
+        # Check that the snoop result was not checked
+        self.mock_get_snoop.assert_not_called()
+
+        # Assert state transition: INVALID -> MODIFIED
+        self.assertEqual(next_state, MESIState.MODIFIED)
+
+    def test_processor_write_invalid_and_hit(self):
```

Comment:
@reecewayt did you have any opinions on how we want to address the different snoop result cases for the processor read / write tests?

Reply:
@nkanderson I'm still unsure how to address this part. I understand that snoop results don't affect the result here, so I guess why test them? I can remove them if you want.

Reply:
Yeah, I'd say for consistency if we're not going to test all the snoop results for all the cases, we shouldn't do it for just one unless there's some reason to, which we'd want to include in a comment. Since I don't think we're inclined to add all the snoop results, we should consolidate these into a single `test_processor_write_invalid`

Reply:
@nkanderson Alright I removed these tests and added `test_processor_write_invalid`